File: 2021/day14.py
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day14.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])

from utils import parsefile, parse

start, *data = parsefile(file_name, [str, 1, [[str, " -> "], "\n"], "\n\n"])

rules = {}
for rule in data[0]:
	rules[rule[0]] = rule[1]

def part1():
	s = start
	
	for it in range(10):
		new_s = ""
		for i in range(1, len(s)):
			new_s += s[i-1]
			new_s += rules[s[i-1:i+1]]
			
		new_s += s[-1]
		s = new_s

	count = {chr(v):0 for v in range(ord("A"), ord("Z")+1)}
	for c in s:
		count[c] += 1
		
	min_count = min([v for v in count.values() if v > 0])
	max_count = max(count.values())
	return max_count - min_count

mem = {}
rules_int = {(ord(k[0])-65, ord(k[1])-65):ord(v)-65 for k, v in rules.items()}
zeroes = [0]*26

"""
based on this
20 is original length
39 is length after transform
split into 2 halves to stop memory overload
each tab represents a iteration
use memorisation or it will take forever
20 -> 39 -> 20, 20
	20 -> 39 -> 20, 20
		20 -> 39 -> 20, 20
		20 -> 39 -> 20, 20
	20 -> 39 -> 20, 20
		20 -> 39 -> 20, 20
		20 -> 39 -> 20, 20
"""


def rec(s, depth, max_depth, mid):
	if depth >= max_depth:
		c = zeroes.copy()
		for v in s[:-1]:
			c[v] += 1
		return c

	depth += 1

	if (tuple(s), depth) in mem:
		return mem[(tuple(s), depth)]

	new_s = [0] * (2*mid-1)
	new_s_i = 0
	for i in range(1, mid):
		new_s[new_s_i] = s[i - 1]
		new_s[new_s_i+1] = rules_int[tuple(s[i-1:i+1])]
		new_s_i += 2

	new_s[-1] = s[-1]

	count = zeroes.copy()

	c1 = rec(new_s[:mid], depth, max_depth, mid)
	c2 = rec(new_s[mid-1:], depth, max_depth, mid)

	for i in range(26):
		count[i] += c1[i] + c2[i]

	mem[(tuple(s), depth)] = count

	return count


def part2():
	mid = len(start)

	start_int = [ord(c) - 65 for c in start]
	iterations = 40
	count = rec(start_int, 0, iterations, mid)

	count[start_int[-1]] += 1

	min_count = min([v for v in count if v > 0])
	max_count = max(count)
	return max_count - min_count


p1()
p2()

File: cpp/2021/day14.cpp
#include "../aocHelper.h"

class Day14 : public BaseDay
{
public:
	Day14() : BaseDay("14") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		map<pair<char, char>, char> rules;
		string poly;

		while (*input != '\n')
		{
			poly += *input;
			input++;
		}

		input += 2; // skip '\n\n'

		while (*input != '\0')
		{
			char c1 = *input;
			input++;
			char c2 = *input;
			input++;

			input += 4; // skip ' -> '

			char c3 = *input;
			input++;

			rules.insert({ { c1, c2}, c3 });

			input++; // skip '\n'
		}

		/*
		based on this
		20 is original length
		39 is length after 1 transform
		split into 2 halves (0-19 and 19-38) to stop memory overload
		each tab represents a iteration
		use memorisation or it will take forever
		20 -> 39 -> 20, 20
			20 -> 39 -> 20, 20
				20 -> 39 -> 20, 20
				20 -> 39 -> 20, 20
			20 -> 39 -> 20, 20
				20 -> 39 -> 20, 20
				20 -> 39 -> 20, 20
		*/

		map<pair<int, string>, array<long long, 26>> mem;

		function<array<long long, 26>(string& str, int, int, int)> step;
		step = [&rules, &mem, &step](string& str, int depth, int max_depth, int mid)
		{
			if (depth >= max_depth)
			{
				array<long long, 26> count{};

				for (int i = 0; i < str.size() - 1; i++)
				{
					//char c = str[i];
					//int a = str[i] - 'A';
					count[str[i] - 'A']++;
				}
				return count;
			}

			depth++;

			if (mem.contains({ depth, str }))
			{
				return mem[{ depth, str }];
			}

			string new_str;
			new_str.resize(2 * mid - 1, '-');

			int new_str_i = 0;
			for (int i = 1; i < mid; i++)
			{
				new_str[new_str_i] = str[i - 1];
				new_str[new_str_i + 1] = rules[{ str[i - 1], str[i] }];
				new_str_i += 2;
			}

			new_str.back() = str.back();

			array<long long, 26> count{};

			string s1{ new_str.begin(), new_str.begin() + mid };
			string s2{ new_str.begin() + mid - 1, new_str.end() };

			auto c1 = step(s1, depth, max_depth, mid);
			auto c2 = step(s2, depth, max_depth, mid);

			for (int i = 0; i < 26; i++)
			{
				count[i] += c1[i] + c2[i];
			}

			mem[{depth, str}] = count;

			return count;
		};

		// part 1
		auto count_p1 = step(poly, 0, 10, poly.size());

		count_p1[poly.back() - 'A']++;

		{
			int min_i = 0;
			int min_v = INT_MAX;

			int max_i = 0;
			int max_v = 0;

			for (int i = 0; i < 26; i++)
			{
				if (count_p1[i] > 0 && count_p1[i] < min_v)
				{
					min_i = i;
					min_v = count_p1[i];
				}

				if (count_p1[i] > max_v)
				{
					max_i = i;
					max_v = count_p1[i];
				}
			}

			part1 = max_v - min_v;
		}

		// part 2
		mem.clear();

		auto count_p2 = step(poly, 0, 40, poly.size());

		count_p2[poly.back() - 'A']++;

		{
			long long min_i = 0;
			long long min_v = LLONG_MAX;

			long long max_i = 0;
			long long max_v = 0;

			for (int i = 0; i < 26; i++)
			{
				if (count_p2[i] > 0LL && count_p2[i] < min_v)
				{
					min_i = i;
					min_v = count_p2[i];
				}

				if (count_p2[i] > max_v)
				{
					max_i = i;
					max_v = count_p2[i];
				}
			}

			part2 = max_v - min_v;
		}

		return { part1, part2 };
	}
};
